R/read.R

```diff
@@ -289,16 +298,57 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
     mat
 }
 
+.read_convert <- function(file, path, recursive=FALSE) {
+    .convert_special_types(rhdf5::h5read(file, path),
+                           recursive=recursive)
+}
+
+.convert_special_types <- function(obj, recursive=FALSE) {
+    if (identical(names(obj), c('categories', 'codes'))) {
+        codes <- obj[['codes']] + 1
+        codes[codes == 0] <- NA
+        levels <- obj[['categories']]
+        obj <- factor(levels[codes], levels=levels)
+        return(obj)
+    }
+
+    if (identical(names(obj), c('mask', 'values'))) {
+        mask <- as.logical(obj[['mask']]) # convert enum to bool
+        obj <- obj[['values']]
+        obj[mask] <- NA
+    }
+
+    if (is.factor(obj) && identical(levels(obj), c('FALSE', 'TRUE'))) {
+        obj <- as.logical(obj)
+    }
+
+    if (recursive && is.list(obj)) {
+        obj <- lapply(
+            obj, function(x) .convert_special_types(x, recursive=TRUE)
+        )
+    }
+
+    if (is.list(obj) && !is.null(names(obj))) {
+        names(obj) <- make.names(names(obj))
+    }
+
+    obj
+}
+
 #' @importFrom S4Vectors DataFrame
 .read_dim_data <- function(file, path, fields) {
     col_names <- setdiff(names(fields), c("__categories", "_index"))
     out_cols <- list()
     for (col_name in col_names) {
-        out_cols[[col_name]] <- as.vector(
-            rhdf5::h5read(file, file.path(path, col_name))
-        )
+        out_cols[[col_name]] <- .read_convert(file, file.path(path, col_name),
+                                              recursive=FALSE)
```

Comment:
Ah, I see we use `recursive=FALSE` here, can probably ignore the earlier comment then